File: app/src/main/java/danbroid/busapp/activities/MainActivity.java
package danbroid.busapp.activities;

import android.app.SearchManager;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.Toast;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.App;
import org.androidannotations.annotations.Background;
import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.OptionsItem;
import org.androidannotations.annotations.OptionsMenu;
import org.androidannotations.annotations.UiThread;
import org.androidannotations.annotations.ViewById;

import danbroid.busapp.BuildConfig;
import danbroid.busapp.BusApp;
import danbroid.busapp.R;
import danbroid.busapp.db.BusStopDB;
import danbroid.busapp.db.model.BusStop;
import danbroid.busapp.interfaces.HandlesBackButton;
import danbroid.busapp.interfaces.MainView;
import danbroid.busapp.interfaces.SwipeRefreshable;
import danbroid.busapp.ui.AboutDialogHelper;
import danbroid.busapp.ui.MetlinkLiveInfo;
import danbroid.busapp.ui.RecentStops;
import danbroid.busapp.ui.WebBrowser;
import danbroid.busapp.ui.views.SwipeRefresh;
import danbroid.busapp.util.HelpCodes;
import danbroid.touchprompt.TouchPrompt;

@EActivity(R.layout.main_activity)
@OptionsMenu(R.menu.main_activity)
public class MainActivity extends AppCompatActivity implements MainView {

  private static final org.slf4j.Logger log = org.slf4j.LoggerFactory.getLogger(MainActivity.class);


  @App
  BusApp app;

  @Bean
  BusStopDB busStopDB;

  @ViewById(R.id.progress_bar)
  View progressBar;

  private long backPressedAt = 0;

  @ViewById(R.id.toolbar)
  Toolbar toolbar;

  @ViewById(R.id.swipe_refresh)
  SwipeRefresh swipeRefresh;


  @Override
  protected void onNewIntent(Intent intent) {
    super.onNewIntent(intent);
    log.info("onNewIntent() action:{} stop_code:{}", intent.getAction(), intent.getStringExtra
        ("stop_code"));

    processIntent(intent);
  }


  @AfterViews
  void init() {


    setSupportActionBar(toolbar);

    processIntent(getIntent());


    swipeRefresh.setOnRefreshListener(() -> {
      Fragment contentView = getContentView();
      if (contentView instanceof SwipeRefreshable) {
        ((SwipeRefreshable) contentView).refresh();
      }
    });

  }

  protected void processIntent(Intent intent) {
    log.info("processIntent(): action:{}", intent.getAction());

    String action = intent.getAction();

    Bundle extras = intent.getExtras();

    if (BuildConfig.DEBUG && extras != null) {
      for (String extra : extras.keySet()) {
        log.trace("extra: {}:{}", extra, extras.get(extra));
      }
    }


    if (action == null) {
    } else if (action.equals(Intent.ACTION_VIEW)) {
      selectStop(intent.getStringExtra(SplashActivity.INTENT_EXTRA_STOP_CODE));
      return;
    } else if (action.equals(Intent.ACTION_SEARCH)) {
      Uri data = intent.getData();
      if (data != null) {
        log.info("Intent.ACTION_SEARCH with: {}", data);

        selectStop(Long.parseLong(data.getLastPathSegment()));

      } else {
        String query = intent.getStringExtra(SearchManager.QUERY);
        log.trace("Intent.ACTION_SEARCH with query: {}", query);
        //not supported

      }
      return;
    }

    if (getContentView() == null) {
      //initialise the main content with the recent stops list
      showRecentStops();
    }

    showHelp();
  }


  protected boolean navigateBack() {
    Fragment contentView = getContentView();

    if (contentView instanceof HandlesBackButton) {
      if (((HandlesBackButton) contentView).onBackButton()) {
        return true;
      }
    }

    FragmentManager fm = getSupportFragmentManager();
    if (fm.getBackStackEntryCount() > 0) {
      fm.popBackStack();
      return true;
    }

    if (!(getContentView() instanceof RecentStops)) {
      showRecentStops();
      return true;
    }

    return false;
  }

  @OptionsItem(android.R.id.home)
  @Override
  public void onBackPressed() {

    if (navigateBack()) return;

    long time = System.currentTimeMillis();

    if (time - backPressedAt < 2000) {
      //back button pressed twice within 2 seconds
      onActionExit();
      return;
    }

    backPressedAt = time;
    Toast.makeText(this, getString(R.string.msg_press_back_to_quit), Toast.LENGTH_SHORT).show();
  }


  void setContentView(Fragment contentView) {
    setContentView(contentView, false);
  }

  public void setContentView(Fragment contentView, boolean addToBackStack) {
    log.trace("setContentView() {}", contentView);


    TouchPrompt.closeChain();
    progressBar.setVisibility(View.GONE);


    setSwipeRefreshEnabled(contentView instanceof SwipeRefreshable);


    FragmentTransaction ft = getSupportFragmentManager().beginTransaction()
        .replace(R.id.fragment_container, contentView);

    if (addToBackStack) ft.addToBackStack(null);
    ft.commit();

    getSupportActionBar().setDisplayHomeAsUpEnabled(!(contentView instanceof RecentStops));
  }

  public Fragment getContentView() {
    return getSupportFragmentManager().findFragmentById(R.id.fragment_container);
  }

  @UiThread
  @Override
  public void setSwipeRefreshEnabled(boolean enabled) {
    swipeRefresh.setEnabled(enabled);
  }


  @UiThread
  @Override
  public void setRefreshing(boolean refreshing) {
    swipeRefresh.setRefreshing(refreshing);
  }

  public void showLiveInfo(BusStop stop) {
    log.info("showLiveInfo(): {} ", stop);
    setContentView(MetlinkLiveInfo.newInstance(stop));

    busStopDB.incrementAccessCount(stop);
  }


  @UiThread
  public void selectStop(BusStop stop) {
    log.trace("selectStop(): {}", stop);
    app.setCurrentStop(stop);
    showLiveInfo(stop);
  }

  @Background
  public void selectStop(String stopCode) {
    log.trace("selectStop(): code: {}", stopCode);
    BusStop stop = busStopDB.getBusStop(stopCode);
    if (stop != null)
      selectStop(stop);
  }

  @Background
  public void selectStop(long stopID) {
    selectStop(busStopDB.getBusStop(stopID));
  }


  private static final String PREF_ABOUT_DIALOG_VERSION = BusApp.class
      .getName() + ".aboutDialogVersion";
  private static final int ABOUT_DIALOG_VERSION = 1;


  /**
   * Displays the about dialog and then displays introductory help prompts
   */
  public void showHelp() {
    log.trace("showHelp()");

    int aboutVersion = getPrefs().getInt(PREF_ABOUT_DIALOG_VERSION, 0);
    log.warn("showHelp() aboutVersion: " + aboutVersion);
    if (aboutVersion != ABOUT_DIALOG_VERSION) {
      getPrefs().edit().putInt(PREF_ABOUT_DIALOG_VERSION, ABOUT_DIALOG_VERSION).apply();
      new AboutDialogHelper(this).showAbout(false, () -> showPrompts());
    } else {
      showPrompts();
    }

  }

  /**
   * Displays introductory help prompts
   */
  private void showPrompts() {

    //TODO: move messages to strings.xml
    new TouchPrompt(this)
        .setTarget(R.id.menu_search)
        .setPrimaryText("Search Button")
        .setShortDelay()
        .setSecondaryText("Click here to search for a stop.\nYou can search using the name of the stop or via its code.")
        .setSingleShotID(HelpCodes.MAIN_SEARCH_BUTTON)
        .show();
    new TouchPrompt(this)
        .setTarget(R.id.menu_map)
        .setShortDelay()
        .setPrimaryText("Map Button")
        .setSecondaryText("Click here to find a stop using a map")
        .setSingleShotID(HelpCodes.MAIN_MAP_BUTTON)
        .show();
  }

  public SharedPreferences getPrefs() {
    return app.getPrefs();
  }

  @OptionsItem(R.id.menu_about)
  public void onActionAbout() {
    log.debug("onActionAbout()");
    new AboutDialogHelper(this).showAbout(true, null);
  }


  @OptionsItem(R.id.menu_exit)
  public void onActionExit() {
    log.info("onActionExit();");
    finish();
  }

  /**
   * Displays the google map view
   */
  @OptionsItem(R.id.menu_map)
  public void showMap() {
    log.debug("showMap();");
    startActivity(new Intent(this, MapActivity_.class));
  }

  /**
   * Display the recent sstops history list
   */
  public void showRecentStops() {
    log.info("showRecentStops()");
    setContentView(RecentStops.getInstance());
  }

  /**
   * Displays a webpage
   *
   * @param url
   */
  @Override
  public void showWebBrowser(String url) {
    log.info("showWebBrowser() :{}", url);
    setContentView(WebBrowser.getInstance(url), true);
  }

}

File: app/src/main/java/danbroid/busapp/activities/MainActivity.kt
package danbroid.busapp.activities


import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Bundle
import android.os.StrictMode
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.content.res.ResourcesCompat
import androidx.core.view.ViewCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.lifecycleScope
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.setupActionBarWithNavController
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import danbroid.busapp.*
import danbroid.busapp.databinding.ActivityMainBinding
import danbroid.busapp.models.AppModel
import danbroid.busapp.ui.AboutDialogHelper
import danbroid.busapp.ui.BusStopSearchView
import danbroid.touchprompt.TouchPrompt
import danbroid.touchprompt.touchPrompt
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory


class MainActivity : AppCompatActivity(), MainActivityInterface {

  protected val navHostFragment: NavHostFragment
    get() = supportFragmentManager.findFragmentById(R.id.nav_host_fragment) as NavHostFragment

  private val navController: NavController
    get() = navHostFragment.navController

  internal lateinit var searchView: BusStopSearchView

  override val appModel by viewModels<AppModel>()

  private lateinit var binding: ActivityMainBinding

  override fun onSupportNavigateUp() = navController.navigateUp() || super.onSupportNavigateUp()

  override fun onCreate(savedInstanceState: Bundle?) {
    log.debug("onCreate() state: $savedInstanceState")
    setTheme(R.style.AppTheme_NoActionBar)
    // AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
    super.onCreate(savedInstanceState)

    val veryStrict = false
    if (BuildConfig.DEBUG && veryStrict) {
      StrictMode.setThreadPolicy(
        StrictMode.ThreadPolicy.Builder()
          .detectCustomSlowCalls()
          .detectDiskWrites()
          .detectNetwork()
          .penaltyLog()
          .build()
      )
      StrictMode.setVmPolicy(
        StrictMode.VmPolicy.Builder()
          .detectLeakedSqlLiteObjects()
          .detectLeakedClosableObjects()
          .penaltyLog()
          .penaltyDeath()
          .build()
      )
    }

    binding = ActivityMainBinding.inflate(layoutInflater).also {
      setContentView(it.root)
    }


    setSupportActionBar(binding.toolbar)

    val navController = navHostFragment.navController

    navController.graph = navController.createBusAppNavGraph(this)


    setupActionBarWithNavController(navController)


    var appBarBackground: Drawable? = null
    var appBarElevation: Float = -1.0f
    var layoutParams: CoordinatorLayout.LayoutParams? = null
    var appBarBehavior: CoordinatorLayout.Behavior<*>? = null

    navController.addOnDestinationChangedListener { controller, destination, arguments ->
      log.info("onDestinationChanged() $destination arg: $arguments")
      if (appBarBackground == null) {
        appBarBackground = binding.appbar.background
        appBarElevation = ViewCompat.getElevation(binding.appbar)
        layoutParams =
          findViewById<View>(R.id.nav_host_fragment).layoutParams as CoordinatorLayout.LayoutParams
        appBarBehavior = layoutParams!!.behavior
      }
      when (destination.id) {
        NavGraph.dest.map -> {
          //binding.appbar.alpha = 0.5f
          binding.appbar.setBackgroundColor(
            ResourcesCompat.getColor(
              resources,
              R.color.colorPrimaryTransparent,
              null
            )
          )
          layoutParams?.behavior = null
          ViewCompat.setElevation(binding.appbar, 0f)
        }
        else -> {
          //binding.appbar.alpha = 1f
          binding.appbar.setBackgroundColor(
            ResourcesCompat.getColor(
              resources,
              R.color.colorPrimary,
              null
            )
          )
          ViewCompat.setElevation(binding.appbar, appBarElevation)
          layoutParams?.behavior = appBarBehavior

        }
      }
      /*    if (appBarBackground == null){
            appBarBackground = binding.appbar.background
            appBarElevation = ViewCompat.getElevation(binding.appbar)
            layoutParams = findViewById<View>(R.id.nav_host_fragment).layoutParams as CoordinatorLayout.LayoutParams
            appBarBehavior = layoutParams!!.behavior
          }
          when (destination.id) {
            NavGraph.dest.map -> {
              binding.toolbar.setTitleTextColor(Color.BLUE)
              layoutParams?.behavior = null
              ViewCompat.setElevation(binding.appbar, 0f)
              binding.appbar.background = null
            }
            else -> {
              layoutParams?.behavior = appBarBehavior
              ViewCompat.setElevation(binding.appbar, appBarElevation)
              binding.appbar.background = appBarBackground
            }

          }*/
    }

//    var toolbarAnim: ViewPropertyAnimator? = null

    closeSearchView()
    appModel.errorMessage.value = null


    // supportActionBar!!.setDisplayHomeAsUpEnabled(supportFragmentManager.backStackEntryCount > 1)

/*    appModel.searchBusy.observe(this, { busy ->

      *//*findViewById<SwipeRefreshLayout>(R.id.swipe_refresh_layout)?.let { swipeLayout ->
        swipeLayout.isRefreshing = busy
      }*//*

      if (busy) {
        binding.progressBar.visibility = View.VISIBLE
        appModel.errorMessage.value = null
      } else {
        binding.progressBar.visibility = View.GONE
      }
    })*/

    appModel.errorMessage.observe(this, Observer(::showError))

    appModel.results.observe(this, Observer
    {
      searchView.results = it
    })

    searchView = BusStopSearchView(this)

    searchView.queryHint = getString(R.string.msg_search_button)

/*  searchView.onActionViewExpandedListener = {
    if (!it) binding.fragment.requestFocus()
  }*/

    searchView.onStopSelectListener = {
      addToHistoryAndShowDepartures(it.code)
    }

    searchView.setOnQueryTextListener(
      object : SearchView.OnQueryTextListener {

        override fun onQueryTextSubmit(query: String): Boolean = false

        override fun onQueryTextChange(query: String): Boolean {
          appModel.performSearch(query)
          return true
        }
      })

    processIntent(intent)

    //if (BuildConfig.DEBUG) TouchPrompt.clearPrefs(this)

    AboutDialogHelper(this).showAbout {
      touchPrompt(HelpCodes.SEARCH_BUTTON) {
        setPrimaryText(R.string.lbl_search_button)
        setSecondaryText(R.string.msg_search_button)
        initialDelay = 1000
        targetID = R.id.action_search
        onDismissed = {
          touchPrompt(HelpCodes.MAP_BUTTON) {
            initialDelay = 500
            setPrimaryText(R.string.lbl_map_button)
            setSecondaryText(R.string.msg_map_button)
            targetID = R.id.action_map
          }
        }
      }
    }
  }

/*  fun showDepartures(stopCode: String) =
    lifecycleScope.launch {
      log.info("showDepartures() $stopCode")
      metlinkOld.stopInfo(stopCode).also {
        showDepartures(it)
      }
    }.invokeOnCompletion {

      if (it != null) {
        if (it !is CancellationException)
          appModel.errorMessage.postValue("Network error: ${it.message}")
      }
    }


  fun showDepartures(stop: BusStop) {
    appModel.currentStop.value = stop
    //setContent(DeparturesFragment(), addToBackStack = addToBackStack)
  }*/


  override fun showBrowser(url: String) {
    startActivity(
      Intent(Intent.ACTION_VIEW)
        .setData(Uri.parse(url))
    )

//    setContent(BrowserFragment.newInstance(url))
  }

  override fun showMap(stopCode: String?) =
    navController.showMap(stopCode)


  override fun setTitle(name: CharSequence) {
    binding.toolbar.title = name
  }


  //private var errorDialog: AlertDialog? = null
  private var alertDialog: androidx.appcompat.app.AlertDialog? = null

  private fun showError(msg: String?) {
    log.info("showError(): $msg")
    alertDialog?.cancel().also {
      alertDialog = null
    }

    msg ?: return

    findViewById<View>(R.id.progress_bar).visibility = View.GONE
    alertDialog = MaterialAlertDialogBuilder(this).apply {
      setTitle(R.string.lbl_alert)
      setMessage(msg)
      setIcon(R.drawable.ic_warning_small)
      setPositiveButton(android.R.string.ok, null)
    }.show()
  }

  fun processIntent(intent: Intent) {
    log.info("processIntent(): $intent")
    intent.extras?.getString(INTENT_EXTRA_STOP_CODE)?.also { code ->
      addToHistoryAndShowDepartures(code)
    }
  }

  override fun addToHistoryAndShowDepartures(code: String) {
    log.info("addToHistoryAndShowDepartures() $code")
    lifecycleScope.launch(Dispatchers.IO) {
      appModel.addStop(code).also {
        log.info("showing departures for $code")
        withContext(Dispatchers.Main) {
          navController.showDepartures(code)
        }
      }
    }.invokeOnCompletion {
      if (it != null)
        log.error(it.message, it)
    }
  }

  override val context: Context = this

  override fun onNewIntent(intent: Intent) {
    super.onNewIntent(intent)
    processIntent(intent)
  }

  private var lastTimeBackPressed = 0L

  override fun onBackPressed() {
    log.debug("onBackPressed back count: ${supportFragmentManager.backStackEntryCount}")

    if (searchView.closeSearchView()) return

    if (navController.navigateUp()) return
/*
  currentContent?.let {
    it is SupportsBackButton && it.onBackPressed() && return
  }
*/


/*
    if (supportFragmentManager.backStackEntryCount > 0) {
      supportFragmentManager.popBackStack()
      return
    }*/

    val now = System.currentTimeMillis()
    if (now - lastTimeBackPressed < 3000) {
      finish()
      return
    }
    lastTimeBackPressed = now
    Toast.makeText(this, R.string.msg_press_back_again_to_quit, Toast.LENGTH_SHORT).show()

  }

  private var menu: Menu? = null

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    this.menu = menu

    menuInflater.inflate(R.menu.main_activity, menu)
    super.onCreateOptionsMenu(menu)

    if (BuildConfig.DEBUG) {

      menu.add("Show test stop").setOnMenuItemClickListener {
        addToHistoryAndShowDepartures("WELL")
        true
      }

      menu.add("Test ERror").setOnMenuItemClickListener {
        appModel.errorMessage.value = "Something bad happened!"
        true
      }

      menu.add("Menu Prompt").setOnMenuItemClickListener {
        touchPrompt {
          primaryText = "Menu Prompt"
          setTargetPosition(100f, 200f)
        }
        true
      }

    }

    return true
  }


  override fun onOptionsItemSelected(item: MenuItem): Boolean =
    when (item.itemId) {
      android.R.id.home -> {
        onBackPressed()
        true
      }
      R.id.action_about -> {
        AboutDialogHelper(this).showAbout(true, true)
        true
      }
      R.id.action_reset_help -> {
        TouchPrompt.clearPrefs(this)
        val nextIntent = intent
        finish()
        startActivity(nextIntent)
        true
      }
      else -> super.onOptionsItemSelected(item)
    }


  fun closeSearchView() = binding.toolbar.collapseActionView()


}

fun Activity?.displayToolbar(visible: Boolean): Unit =
  (this as AppCompatActivity).run {
    if (visible) {
      findViewById<View>(R.id.toolbar).visibility = View.VISIBLE
    } else {
      findViewById<View>(R.id.toolbar).visibility = View.GONE
    }
  }


val Fragment.mainActivity: MainActivityInterface
  get() = requireActivity() as MainActivityInterface

val Fragment.appModel: AppModel
  get() = mainActivity.appModel

private val log =
  LoggerFactory.getLogger(MainActivity::class.java)


